### PR TITLE
new directive: conflicts

### PR DESCRIPTION
**Pull request by alalazo**
##### Modifications
- [x] implemented a `conflicts` directive as discussed in #2795 and #3123
- [x] coded the conflict shown in #3119 for openblas 
- [x] added unit tests
- [x] added docstrings
- [x] added entry in the docs

### CODE REVIEW

**Comment by alalazo** [comment]
@adamjstewart @davydden @citibeth A preview of the current implementation. Adding the following to `zlib`:
```python
conflicts('~shared', when='%gcc')
conflicts('~shared', when='+pic')
conflicts('~pic', when='%gcc')
```
you'll get the following output for e.g. `spack spec`:
```console
$ spack spec zlib
Input spec
--------------------------------
zlib

Normalized
--------------------------------
zlib

Concretized
--------------------------------
zlib@1.2.10%gcc@4.8+pic+shared arch=linux-ubuntu14-x86_64 

$ spack spec zlib~shared
Input spec
--------------------------------
zlib~shared

Normalized
--------------------------------
zlib~shared

Concretized
--------------------------------
==> Error: Conflicts in concretized spec "zlib@1.2.10%gcc@4.8+pic~shared arch=linux-ubuntu14-x86_64 /qyf6e5l"

List of matching conflicts:

1. "~shared" conflicts with "zlib%gcc" in spec "zlib@1.2.10%gcc@4.8+pic~shared arch=linux-ubuntu14-x86_64"
2. "~shared" conflicts with "zlib+pic" in spec "zlib@1.2.10%gcc@4.8+pic~shared arch=linux-ubuntu14-x86_64"

$ spack spec zlib~shared~pic
Input spec
--------------------------------
zlib~pic~shared

Normalized
--------------------------------
zlib~pic~shared

Concretized
--------------------------------
==> Error: Conflicts in concretized spec "zlib@1.2.10%gcc@4.8~pic~shared arch=linux-ubuntu14-x86_64 /vqct3p4"

List of matching conflicts:

1. "~shared" conflicts with "zlib%gcc" in spec "zlib@1.2.10%gcc@4.8~pic~shared arch=linux-ubuntu14-x86_64"
2. "~pic" conflicts with "zlib%gcc" in spec "zlib@1.2.10%gcc@4.8~pic~shared arch=linux-ubuntu14-x86_64"

$ spack spec hdf5 ^ zlib~shared
Input spec
--------------------------------
hdf5
    ^zlib~shared

Normalized
--------------------------------
hdf5
    ^zlib@1.1.2:~shared

Concretized
--------------------------------
==> Error: Conflicts in concretized spec "hdf5@1.10.0-patch1%gcc@4.8+cxx~debug+fortran+mpi+pic+shared~szip~threadsafe arch=linux-ubuntu14-x86_64 /ec5oonm"

List of matching conflicts:

1. "~shared" conflicts with "zlib%gcc" in spec "zlib@1.2.10%gcc@4.8+pic~shared arch=linux-ubuntu14-x86_64"
2. "~shared" conflicts with "zlib+pic" in spec "zlib@1.2.10%gcc@4.8+pic~shared arch=linux-ubuntu14-x86_64"
```

**Comment by adamjstewart** [comment]
@alalazo What about:
```python
conflicts('%pgi')
```
or
```python
conflicts(when='%nag')
```
a lot of software can't be built at all with certain compilers.

Ideally, not necessarily in this PR, a conflict should try to be resolved automatically. If I want to build a stack of software with PGI or NAG and a single package conflicts, instead of crashing, Spack should build that package with the first compiler it finds that does not conflict (usually GCC).

**Comment by alalazo** [comment]
@adamjstewart `conflicts('%pgi')` should already work. Basically if you don't specify `when` it's assumed `when=pkg.name`. I'll add a test for this specific case.

**Comment by citibeth** [comment]
Good work!  Does this solve the problem for #3119?

**Comment by citibeth** [comment]
The original issue is:
```
> # https://github.com/LLNL/spack/issues/3036
> # OpenBLAS (0.2.15:0.2.19) does not support Intel compiler version 16
```
Does the following also work?
```
# See https://github.com/LLNL/spack/issues/3036
conflicts('%intel@16', when='@0.2.15:0.2.19')
```

**Comment by alalazo** [comment]
@citibeth Yes, it should work.

**Comment by davydden** [comment]
> @citibeth Yes, it should work.

would you please add some unit tests for those cases we come up in the discussion?

**Comment by alalazo** [comment]
@davydden I will, definitely! But not now :smile:

**Comment by davydden** [comment]
@alalazo another example of the conflict, can you do the following (example for `slepc` which uses `petsc`):
```
if spec.satisfies('+arpack^petsc+int64'):
      raise RuntimeError('Arpack can not be used with 64bit integers.')
```
i.e. specific variant should not be enabled if a dependecy is build with some variant.

**Comment by alalazo** [comment]
@davydden I think so. It should be:
```
conflicts('+arpack', when='^petsc+int64')
```
If you think it is the case I'll code it in `slepc` and try it out.

**Comment by davydden** [comment]
@alalazo don't bother, I am adding it in https://github.com/LLNL/spack/pull/3223. Just wanted to ask. If this PR will be merged first, i will update mine. Otherwise we would still need to update packages which declare conflicts.

File: lib/spack/spack/spec.py [hunk 1]
```diff
@@ -1551,6 +1551,18 @@ def concretize(self):
...
+        # Now that the spec is concrete we should check if
+        # there are declared conflicts
+        matches = []
+        for x in self.traverse():
+            for conflict_spec, when_list in x.package.conflicts.items():
+                if x.satisfies(conflict_spec):
+                    for when_spec in when_list:
+                        if x.satisfies(when_spec):
+                            matches.append((x, conflict_spec, when_spec))
+        if matches:
```
Comment (alalazo) [inline lib/spack/spack/spec.py]: @davydden The lines above (from the for loop) are the logic behind the `conflicts` directive. Basically with conflicts you store a map that has the conflict as the key, and a list of conditions as values. This allows you to use the same conflict multiple times:
```python
# +foo cannot be used with intel
conflicts('+foo', when='%intel') 
# or when we link with zlib
conflicts('+foo', when='^zlib')
```

If you satisfy the conflict and at least one condition the operation is aborted.

**Comment by alalazo** [comment]
ping

**Comment by alalazo** [comment]
ping

**Comment by alalazo** [comment]
@tgamblin ping

**Comment by tgamblin** [comment]
@alalazo @adamjstewart: it might be worth adding xfail tests for some of the known cases where the concretizer fails to do something smart, like the variant case I mentioned above and some of the painful `+X` cases.  I know it would be helpful for me to have the test cases.

**Pull request merged**